### DISCUSSION TITLE
“What kind of nonsense is this again?” - is happening again with your UPDATE- library code!..

### DISCUSSION THREAD

**Discussion by Knottis** [pschatzmann/ESP32-A2DP#577] [Q&A] [answered]
### ESP32 Wroom-32 connect with I2S to PCM5102

I’ve been so foolish to UPDATE the Arduino libraries I use, specifically ESP32-A2DP and arduino-audio-tools. I know very well that this won’t work after update!

Could you please leave the libraries that work for the ‘**_ESP32 Wroom-32_**’ untouched and instead create drivers that function on the ‘ESP32-A1S’—or whatever it is—in a separate directory or library?

_(Some of the examples use I2S.H and some time ESP_I2S.h)._

**Simply trying this basic code on ‘ESP32 Wroom-32 connect to the PCM5102’- sounds terrible!** 

- There must be a serious issue with BluetoothA2DPSink.

### _I hope you can update/fix the error so that your libraries work with ESP32-Wroom-32 again!_

```c
#include "ESP_I2S.h"
#include "BluetoothA2DPSink.h"

const uint8_t I2S_SCK = 32;     // Audio data bit clock 
const uint8_t I2S_WS = 33;      // Audio data left and right clock 
const uint8_t I2S_SDOUT = 2;    // ESP32 audio data output (to speakers) 
I2SClass i2s;

BluetoothA2DPSink a2dp_sink(i2s);

void setup() {
    i2s.setPins(I2S_SCK, I2S_WS, I2S_SDOUT);
    if (!i2s.begin(I2S_MODE_STD, 44100, I2S_DATA_BIT_WIDTH_16BIT, I2S_SLOT_MODE_STEREO, I2S_STD_SLOT_BOTH)) {
      Serial.println("Failed to initialize I2S!");
      while (1); 
    }

    a2dp_sink.start("Guitar-Summer-2024");
}

void loop() {
}

```

**Comment by pschatzmann** [comment]
I am not aware of any changes from my side that would break your sketch above. Did this sketch really work before ?
You seem to be using the new ESP32 3.0.1 core version now which has been released only 2 weeks ago!

I  don't quite understand that you  were writing that you upgraded the AudioTools, but the sketch above is using the Arduino I2S functionality instead. Did you try with an output sketch using the AudioTools I2SStream ?

Can you also provide a little bit more detail what the exact issue is ?

It does not matter what ESP32 hardware you use: the moving target is the Arduino ESP32 core.

**Comment by pschatzmann** [comment]
I noticed that the I2S Api provided by Arduino does not work properly: So please use the AudioTools!

**Comment by Knottis** [comment]
### Hello

Here’s how this code sounds when it runs on the test bench!.. 

See how _BLT_ gets chopped up and transferred to pcm5102. 

It wasn’t like this before I updated the libraries!..

Audio/Video:

https://github.com/pschatzmann/ESP32-A2DP/assets/35444507/dd24bc32-b6ec-49ae-ac91-d0398d9ef09a

**Reply by pschatzmann** [reply]
Did you use the I2SStream ?

**Reply by Knottis** [reply]
🎶 I only run this test code/program on this ESP32 to find out which component is out of order!... 🎶

```c++
#include "ESP_I2S.h"
#include "BluetoothA2DPSink.h"

const uint8_t I2S_SCK = 32;     // Audio data bit clock 
const uint8_t I2S_WS = 33;      // Audio data left and right clock 
const uint8_t I2S_SDOUT = 2;    // ESP32 audio data output (to speakers) 
I2SClass i2s;

BluetoothA2DPSink a2dp_sink(i2s);

void setup() {
    i2s.setPins(I2S_SCK, I2S_WS, I2S_SDOUT);
    if (!i2s.begin(I2S_MODE_STD, 44100, I2S_DATA_BIT_WIDTH_16BIT, I2S_SLOT_MODE_STEREO, I2S_STD_SLOT_BOTH)) {
      Serial.println("Failed to initialize I2S!");
      while (1); 
    }

    a2dp_sink.start("Guitar-Summer-2024");
}

void loop() {
}

```

**Reply by pschatzmann** [reply]
As I was writing above it is the Arduino ESP32 I2S implementation which is causing the issue and the [AudioTools I2SStream](https://github.com/pschatzmann/ESP32-A2DP?tab=readme-ov-file#defining-pins) should works w/o problems. 

I have never been using this API, so I don't know what's wrong: I would have expected that this just works since it uses the same IDF I2S API that I am also using in the AudioTools. 

Any help will be appreciated!

**Reply by pschatzmann** [reply]
Did you try to investigate the problem ? If you look in the source code the functionality waits for ESP_A2D_CONNECTION_STATE_DISCONNECTED. 
Maybe you never get this ? and you would need to tell me why...

**Reply by pschatzmann** [reply, marked as answer]
I committed a correction to the A2DP library which should correct your issue: The Arduino I2S API should work now properly as well

**Reply by pschatzmann** [reply]
I also increased the default setting for A2DP_I2S_MAX_WRITE_SIZE in config.h

**Comment by Knottis** [comment]
👍Thanks. Now I got the sound back in the guitar case!.. 😉

**Comment by Knottis** [comment]
### _😏One last question!..._
🙏Do you support or when will you support the **ESP 3.0.2 - firmware** !..

**Arduino IDE 2.3.2** => https://www.arduino.cc/en/software
**ESP 3.0.2** => https://github.com/espressif/arduino-esp32

**Reply by pschatzmann** [reply]
There are no relevant changes to 3.0.2 compared to 3.0.1!
So there is nothing to be done from my side...
